### PR TITLE
settings,*: update the settings.Setting API

### LINKED ISSUE #108903: sql: the cluster names are too often marked redactable in logs and errors
**Describe the problem**

Currently the cluster setting name is represented as go `string` and gets redacted away when included in redactable logs and error messages.

**Expected behavior**

When a cluster setting name is included in logs or errors, it should be marked non-redactable.

Preferably this should use a dedicated go type because changing all the call sites of `log.XXX` or `errors.New` would be too cumbersome.

Epic: CRDB-27642

Jira issue: CRDB-30698

### PR DESCRIPTION
**Pull request by knz**
Fixes #108903.
Informs: #108508
Informs: #109046
Epic: CRDB-27642

TLDR: this commit introduces an API distinction between:

- the *name* of a cluster setting used for user-facing UX, including the SQL syntax and error messages.
- the *key* of a cluster setting used to store its value in `system.settings` and organize various data structures.

(In this commit, the name and key remain equivalent to each other; only the interfaces change.)

This change achieves the following:

- it introduces specific Go types for both key and name. The benefit here is to avoid the go `string` type; and force users of the API to consider what data is used through the settings API.

  As a side benefit, it ensures that every setting key or name included in logs and error messages is not redacted away.

- it paves the way for a later change where the name is allowed to diverge from the key. This will allow us to enhance UX without breaking compatibility.

- as a side-benefit, it also marks the "setting origin" attribute
  as non-redactable.

Salient API change, prior to this change:
```go
type Setting interface {
  // Key returns the name of the specific cluster setting.
  Key() string
}
```

After this change:

```go
type Setting interface {
  // InternalKey returns the internal key used to store the setting.
  // To display the name of the setting (eg. in errors etc) or the
  // SET/SHOW CLUSTER SETTING statements, use the Name() method instead.
  //
  // The internal key is suitable for use with:
  // - direct accesses to system.settings.
  // - rangefeed logic associated with system.settings.
  // - (in unit tests only) interchangeably with the name for SET CLUSTER SETTING
  InternalKey() InternalKey

  // Name is the user-visible (display) name of the setting.
  // This is suitable for:
  // - SHOW/SET CLUSTER SETTING.
  // - inclusion in error messages.
  Name() SettingName
}
```

Release note: None

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
Nice! I have some nits that are related to the following questions:
- do we want to apply `sql.defaults` restriction to only keys, only names, or both?
- it seems like in some unit tests we could be using `Name` in more places. Is there a reason why not? It seems like it'd be nice to use `Name` unless it doesn't work because we need `InternalKey` type.

> Reviewed 70 of 70 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-:-Nc5Qhb_BmWabAAtbhqI:b-t8p4ht)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @stevendanna)

___
*[`pkg/cli/gen.go` line 265 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5L4Ei0xLk7Jov89Tn:-Nc5L4Ei0xLk7Jov89To:bewejd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/cli/gen.go#L265)):*
> ```Go
> 				alterRoleLink = `<a href="alter-role.html"><code>ALTER ROLE... SET</code></a>`
> 			}
> 			if strings.Contains(string(name), "sql.defaults") {
> ```

Shouldn't this be `key`? Or perhaps both?
___
*[`pkg/server/admin.go` line 2041 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5Lv-mB-x_5IWaxJgC:-Nc5Lv-mB-x_5IWaxJgD:bp5mht7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/server/admin.go#L2041)):*
> ```Go
> 			for _, k := range settingsKeys {
> 				if slices.Contains(settings.ConsoleKeys(), k) {
> 					newSettingsKeys = append(settingsKeys, k)
> ```

I think this should be `newSettingKeys = append(newSettingKeys, k)`.
___
*[`pkg/server/settingswatcher/settings_watcher.go` line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5MUhM27R2sNnYDrc_:-Nc5MUhM27R2sNnYDrca:bjwuhng) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/server/settingswatcher/settings_watcher.go#L269)):*
> ```Go
> 		}
> 		if setting.Class() != settings.TenantWritable {
> 			log.Warningf(ctx, "ignoring read-only setting %s", settingKey)
> ```

nit: should we use `Name` here?
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5MiGkAkf8o06Nzoid:-Nc5MiGkAkf8o06Nzoie:b-3fwkza) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/server/tenantsettingswatcher/overrides_store.go#L149)):*
> ```Go
> 		after = append(after, setting)
> 	}
> 	// Skip any existing setting for this name.
> ```

nit: perhaps replace "name" with "setting key" or something.
___
*[`pkg/settings/registry.go` line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5NEAj5reQIlM8yqAU:-Nc5NEAj5reQIlM8yqAV:b3vnr1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/settings/registry.go#L320)):*
> ```Go
> // NameToKey returns the key associated with a setting name.
> func NameToKey(name SettingName) (InternalKey, bool) {
> 	// TODO(...): improve this.
> ```

nit: I think our style guide says always include the github username in TODOs even if you're personally not planning to address it.
___
*[`pkg/settings/registry.go` line 349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5NU10-01RRwTZrZzl:-Nc5NU1142q2AKtYbUvm:bhbt9tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/settings/registry.go#L349)):*
> ```Go
> 
> // LookupForReporting returns a Setting by name. Used when a setting is being
> // retrieved for reporting.
> ```

nit: perhaps insert an empty line (for symmetry with the comment below).
___
*[`pkg/sql/show_cluster_setting.go` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5OrKV5-JZzVFGbbjc:-Nc5OrKV5-JZzVFGbbjd:b3szawp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/sql/show_cluster_setting.go#L168)):*
> ```Go
> 	}
> 
> 	if strings.HasPrefix(n.Name, "sql.defaults") {
> ```

nit: perhaps use `name` here too.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go` line 1462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5PKAZ4Z7T8XaJ7WKU:-Nc5PKAZ4Z7T8XaJ7WKV:bl8kojg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go#L1462)):*
> ```Go
> 	}
> 	for _, setting := range settings {
> 		sysDB.Exec(t, fmt.Sprintf("ALTER TENANT ALL SET CLUSTER SETTING %s = $1", setting.InternalKey()), setting.Default()*100)
> ```

nit: why not use `Name` here?
___
*[`pkg/sql/show_test.go` line 1217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5Q9-eENh4l5LMcXL5:-Nc5Q9-eENh4l5LMcXL6:b-3tut8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/sql/show_test.go#L1217)):*
> ```Go
> 
> 	for rows.Next() {
> 		var settingKey, sType, desc string
> ```

nit: technically speaking, shouldn't it be named `settingName`? IIUC `SHOW ALL CLUSTER SETTINGS` uses `crdb_internal.cluster_settings` virtual table which is now updated to return the setting name.

Relatedly, perhaps we want to run the linter on both the key and the name?
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_test.go` line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5Qaou20rWsxAqMI8G:-Nc5Qaou20rWsxAqMI8H:b95uicf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/sql/catalog/schematelemetry/schema_telemetry_test.go#L67)):*
> ```Go
> 
> 	qSet = fmt.Sprintf(`SET CLUSTER SETTING %s = '* * * * *'`,
> 		schematelemetrycontroller.SchemaTelemetryRecurrence.InternalKey())
> ```

nit: this also seems like it could be `Name`.
___
*[`pkg/upgrade/upgrades/ensure_sql_schema_telemetry_schedule_test.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5Qe0B3M9VwlsTSRC9:-Nc5Qe0B3M9VwlsTSRCA:b-ul8jqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/upgrade/upgrades/ensure_sql_schema_telemetry_schedule_test.go#L99)):*
> ```Go
> 		// Check that the schedule can have its recurrence altered.
> 		tdb.Exec(t, fmt.Sprintf(`SET CLUSTER SETTING %s = '* * * * *'`,
> 			schematelemetrycontroller.SchemaTelemetryRecurrence.InternalKey()))
> ```

nit: ditto here and below.

**Review by knz — COMMENTED** [review]
> do we want to apply `sql.defaults` restriction to only keys, only names, or both?

Let's do both.

> it seems like in some unit tests we could be using `Name` in more places. Is there a reason why not?

There's no good reason. Let's do it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-:-Nc6glEq7KQG8SbKX2tu:b-nuzjxa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @stevendanna and @yuzefovich)

___
*[`pkg/cli/gen.go` line 265 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5L4Ei0xLk7Jov89Tn:-Nc6fN7v6wGb8-ZMhwzJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/cli/gen.go#L265)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Shouldn't this be `key`? Or perhaps both?
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 2041 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5Lv-mB-x_5IWaxJgC:-Nc6fS9nBDPOO0C3qsIL:b-k2rwtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/server/admin.go#L2041)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think this should be `newSettingKeys = append(newSettingKeys, k)`.
</blockquote></details>

Good catch. Fixed.
___
*[`pkg/server/settingswatcher/settings_watcher.go` line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5MUhM27R2sNnYDrc_:-Nc6fd648s9e0EXTJWrU:b-oqii3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/server/settingswatcher/settings_watcher.go#L269)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we use `Name` here?
</blockquote></details>

We could, but it's not in scope. I don't feel it's worth performing a lookup.
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5MiGkAkf8o06Nzoid:-Nc6fgVO7iZuOaZU2GUL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/server/tenantsettingswatcher/overrides_store.go#L149)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps replace "name" with "setting key" or something.
</blockquote></details>

Done.
___
*[`pkg/settings/registry.go` line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5NEAj5reQIlM8yqAU:-Nc6fm4FD-s7VPTLOlhF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/settings/registry.go#L320)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I think our style guide says always include the github username in TODOs even if you're personally not planning to address it.
</blockquote></details>

Done.
___
*[`pkg/settings/registry.go` line 349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5NU10-01RRwTZrZzl:-Nc6fu9IE1OY128i9Hvs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/settings/registry.go#L349)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps insert an empty line (for symmetry with the comment below).
</blockquote></details>

Done.
___
*[`pkg/sql/show_cluster_setting.go` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5OrKV5-JZzVFGbbjc:-Nc6g3vI5JIE_gxIRf6-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/sql/show_cluster_setting.go#L168)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps use `name` here too.
</blockquote></details>

Done.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go` line 1462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5PKAZ4Z7T8XaJ7WKU:-Nc6g4vYDmHV5OoTGUNV:b-5sxps5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go#L1462)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why not use `Name` here?
</blockquote></details>

IMHO in tests it doesn't matter as much. But ok - fixed.
___
*[`pkg/sql/show_test.go` line 1217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5Q9-eENh4l5LMcXL5:-Nc6gC1c5T9nf5nPsPeT:bvl5p7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/sql/show_test.go#L1217)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: technically speaking, shouldn't it be named `settingName`? IIUC `SHOW ALL CLUSTER SETTINGS` uses `crdb_internal.cluster_settings` virtual table which is now updated to return the setting name.

Relatedly, perhaps we want to run the linter on both the key and the name?
</blockquote></details>

Actually the linter is right here -- it only needs to operate on names, which is the only thing that's user-facing. So the logic was right from the start.
Just needed to rename settingKey -> settingName here. Done.
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_test.go` line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5Qaou20rWsxAqMI8G:-Nc6geAZ8r7OmC6JzS5N:bwmb33q) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/sql/catalog/schematelemetry/schema_telemetry_test.go#L67)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this also seems like it could be `Name`.
</blockquote></details>

Indeed. Fixed.
___
*[`pkg/upgrade/upgrades/ensure_sql_schema_telemetry_schedule_test.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc5Qe0B3M9VwlsTSRC9:-Nc6gjgS3AvkRBz6SMXb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b6910e6b884222fda54e54027a4a6f06a4dea5b/pkg/upgrade/upgrades/ensure_sql_schema_telemetry_schedule_test.go#L99)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: ditto here and below.
</blockquote></details>

Done.

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> Reviewed 15 of 15 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-:-Nc8nDbz-EplxwwJRdGd:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @stevendanna)

___
*[`pkg/kv/kvpb/api.proto` line 3259 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc8k-Cf60nzLnbheVQd:-Nc8k-Cf60nzLnbheVQe:bboun88) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ce0c2fb1d486c3cbde1b416a2b4047d5eb6658/pkg/kv/kvpb/api.proto#L3259)):*
> ```proto
> }
> 
> // TenantSetting contains the name and value of a tenant setting.
> ```

nit: perhaps `s/name/internal key/`.
___
*[`pkg/server/serverpb/admin.proto` line 620 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc8jiZTBybr9skM849O:-Nc8jiZTBybr9skM849P:b-e4kral) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ce0c2fb1d486c3cbde1b416a2b4047d5eb6658/pkg/server/serverpb/admin.proto#L620)):*
> ```proto
>       bool public = 4;
>       google.protobuf.Timestamp last_updated = 5 [(gogoproto.nullable) = true, (gogoproto.stdtime) = true];
>       string name = 6; // for display purposes.
> ```

nit: this comment is included in generated docs, should we make it a full sentence that starts with the capital?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108902)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[`pkg/kv/kvpb/api.proto` line 3259 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc8k-Cf60nzLnbheVQd:-Nc9QF3a6fBgz6K0kLcM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ce0c2fb1d486c3cbde1b416a2b4047d5eb6658/pkg/kv/kvpb/api.proto#L3259)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps `s/name/internal key/`.
</blockquote></details>

Done.
___
*[`pkg/server/serverpb/admin.proto` line 620 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108902#-Nc8jiZTBybr9skM849O:-Nc9QZRD3JS1tr0urxlP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ce0c2fb1d486c3cbde1b416a2b4047d5eb6658/pkg/server/serverpb/admin.proto#L620)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this comment is included in generated docs, should we make it a full sentence that starts with the capital?
</blockquote></details>

Done.

**Comment by knz** [comment]
TFYR!

bors r=yuzefovich
